File: frontend/src/types/institution/reports.ts
export interface AttendanceReport {
  student_name: string;

  enrollment_number: string;

  department: string;

  program: string;

  batch: string;

  total_classes: number;

  attended: number;

  attendance_percentage: number;

  is_shortage: boolean;
}

export interface SubjectAttendanceReport {
  subject_name: string;

  subject_code: string;

  total_classes: number;

  attended: number;

  attendance_percentage: number;
}

export interface PerformanceReport {
  student_name: string;

  enrollment_number: string;

  department: string;

  program: string;

  total_marks: number;

  avg_marks: number;

  grade: string;

  cgpa: number;
}

export interface FacultyActivityReport {
  faculty_name: string;

  employee_id: string;

  department: string;

  classes_scheduled: number;

  classes_conducted: number;

  attendance_submissions: number;

  subjects_assigned: number;
}

export interface BatchPerformanceReport {
  batch_name: string;

  program_name: string;

  total_students: number;

  avg_attendance: number;

  avg_marks: number;

  top_performer: string;

  shortage_count: number;
}

export interface CertificationReport {
  course_name: string;

  issued: number;

  pending: number;

  revoked: number;
}

export interface StudentProgressReport {
  student_name: string;

  enrollment_number: string;

  current_semester: number;

  overall_attendance: number;

  cgpa: number;

  subjects: Record<string, unknown>[];
}

export interface AtRiskStudent {
  id: number;

  name: string;

  enrollment_number: string;

  attendance: number;

  cgpa: number;

  risk_level: "low" | "medium" | "high";

  risk_reason: string;
}

File: backend/django-service/institution/serializers/reports.py
from rest_framework import serializers


class AttendanceReportSerializer(serializers.Serializer):
    student_name = serializers.CharField()
    enrollment_number = serializers.CharField()
    department = serializers.CharField()
    program = serializers.CharField()
    batch = serializers.CharField()
    total_classes = serializers.IntegerField()
    attended = serializers.IntegerField()
    attendance_percentage = serializers.FloatField()
    is_shortage = serializers.BooleanField()


class SubjectAttendanceSerializer(serializers.Serializer):
    subject_name = serializers.CharField()
    subject_code = serializers.CharField()
    total_classes = serializers.IntegerField()
    attended = serializers.IntegerField()
    attendance_percentage = serializers.FloatField()


class PerformanceReportSerializer(serializers.Serializer):
    student_name = serializers.CharField()
    enrollment_number = serializers.CharField()
    department = serializers.CharField()
    program = serializers.CharField()
    total_marks = serializers.FloatField()
    avg_marks = serializers.FloatField()
    grade = serializers.CharField()
    cgpa = serializers.FloatField()


class FacultyActivityReportSerializer(serializers.Serializer):
    faculty_name = serializers.CharField()
    employee_id = serializers.CharField()
    department = serializers.CharField()
    classes_scheduled = serializers.IntegerField()
    classes_conducted = serializers.IntegerField()
    attendance_submissions = serializers.IntegerField()
    subjects_assigned = serializers.IntegerField()


class BatchPerformanceReportSerializer(serializers.Serializer):
    batch_name = serializers.CharField()
    program_name = serializers.CharField()
    total_students = serializers.IntegerField()
    avg_attendance = serializers.FloatField()
    avg_marks = serializers.FloatField()
    top_performer = serializers.CharField()
    shortage_count = serializers.IntegerField()


class CertificationReportSerializer(serializers.Serializer):
    course_name = serializers.CharField()
    issued = serializers.IntegerField()
    pending = serializers.IntegerField()
    revoked = serializers.IntegerField()


class StudentProgressReportSerializer(serializers.Serializer):
    student_name = serializers.CharField()
    enrollment_number = serializers.CharField()
    current_semester = serializers.IntegerField()
    overall_attendance = serializers.FloatField()
    cgpa = serializers.FloatField()
    subjects = serializers.ListField(child=serializers.DictField())

class AtRiskStudentSerializer(serializers.Serializer):
    RISK_LEVEL_CHOICES = [

        (
            "low",
            "Low",
        ),

        (
            "medium",
            "Medium",
        ),

        (
            "high",
            "High",
        ),

    ]
    id = serializers.IntegerField()
    name = serializers.CharField()
    enrollment_number = serializers.CharField()
    attendance = serializers.FloatField()
    cgpa = serializers.FloatField()
    risk_level = (
        serializers.ChoiceField(

            choices=
                RISK_LEVEL_CHOICES

        )
    )
    risk_reason = serializers.CharField()